### PR TITLE
HPCC-15292 Redis Plugin - make connection caching optional

### PR DESCRIPTION
**Pull request by jamienoss**
Signed-off-by: James Noss james.noss@lexisnexis.com

### CODE REVIEW

**Comment by jamienoss** [comment]
@ghalliday please review.
There is no clean/trivial method for adding a specific runtime regression test for this.

File: plugins/redis/redis.cpp [hunk 1]
```diff
@@ -373,35 +382,53 @@ void Connection::readReplyAndAssertWithCmdMsg(Reply * reply, const char * msg, c
...
     assertOnErrorWithCmdMsg(reply->query(), msg, key);
 }
-Connection * Connection::createConnection(ICodeContext * ctx,  Connection * & _cachedConnection, const char * options, int _database, const char * password, unsigned _timeout)
+Connection * Connection::createConnection(ICodeContext * ctx,  Connection * & _cachedConnection, const char * options, int _database, const char * password, unsigned _timeout, bool cacheConnections)
 {
-#if HIREDIS_VERSION_OK
-    if (!_cachedConnection)
+#if HIREDIS_VERSION_OK_FOR_CACHING
+    //Fetch connection caching level
+    if (!connectionCachingLevelChecked)//test to guard against unnecessary critical section
```
Comment (ghalliday) [inline plugins/redis/redis.cpp]: The code is fine, but it adds unnecessary complication to a function that is not time critical.

File: plugins/redis/redis.cpp [hunk 2]
```diff
@@ -41,11 +42,19 @@ namespace RedisPlugin {
...
-static __thread ThreadTermFunc threadHookChain = NULL;
+
+static __thread Connection * cachedConnection = nullptr;
+static __thread Connection * cachedPubConnection = nullptr;//database should always = 0
+
+#if HIREDIS_VERSION_OK_FOR_CACHING
+static CriticalSection critsec;
+static __thread ThreadTermFunc threadHookChain = nullptr;
 static __thread bool threadHooked = false;
+#define NO_CONNECTION_CACHING 0
```
Comment (ghalliday) [inline plugins/redis/redis.cpp]: minor style: More conventional to use a enum instead of #defines.

**Comment by ghalliday** [comment]
@jamienoss a couple of minor comments, but will merge as-is

**Comment by ghalliday** [comment]
Is there a JIRA open to update this to use the plugin options from the config manager once that functionality is available?

**Comment by jamienoss** [comment]
@ghalliday I will open one now...

**Comment by jamienoss** [comment]
@ghalliday https://track.hpccsystems.com/browse/HPCC-15510

**Pull request merged**